Pack multiple vertex and index arrays together into growable buffers.

pcwalton opened the pull request:
This commit uses the [`offset-allocator`] crate to combine vertex and index arrays from different meshes into single buffers. Since the primary source of `wgpu` overhead is from validation and synchronization when switching buffers, this significantly improves Bevy's rendering performance on many scenes.

This patch is a more flexible version of #13218, which also used slabs. Unlike #13218, which used slabs of a fixed size, this commit implements slabs that start small and can grow. In addition to reducing memory usage, supporting slab growth reduces the number of vertex and index buffer switches that need to happen during rendering, leading to improved performance. To prevent pathological fragmentation behavior, slabs are capped to a maximum size, and mesh arrays that are too large get their own dedicated slabs.

As an additional improvement over #13218, this commit allows the application to customize all allocator heuristics. The `MeshAllocatorSettings` resource contains values that adjust the minimum and maximum slab sizes, the cutoff point at which meshes get their own dedicated slabs, and the rate at which slabs grow. Hopefully-sensible defaults have been chosen for each value.

Unfortunately, WebGL 2 doesn't support the *base vertex* feature, which is necessary to pack vertex arrays from different meshes into the same buffer. `wgpu` represents this restriction as the downlevel flag `BASE_VERTEX`. This patch detects that bit and ensures that all vertex buffers get dedicated slabs on that platform. Even on WebGL 2, though, we can combine all *index* arrays into single buffers to reduce buffer changes, and we do so.

The following measurements are on Bistro:

Overall frame time improves from 8.74 ms to 5.53 ms (1.58x speedup):
![Screenshot 2024-07-09 163521](https://github.com/bevyengine/bevy/assets/157897/5d83c824-c0ee-434c-bbaf-218ff7212c48)

Render system time improves from 6.57 ms to 3.54 ms (1.86x speedup):
![Screenshot 2024-07-09 163559](https://github.com/bevyengine/bevy/assets/157897/d94e2273-c3a0-496a-9f88-20d394129610)

Opaque pass time improves from 4.64 ms to 2.33 ms (1.99x speedup):
![Screenshot 2024-07-09 163536](https://github.com/bevyengine/bevy/assets/157897/e4ef6e48-d60e-44ae-9a71-b9a731c99d9a)

## Migration Guide

### Changed

* Vertex and index buffers for meshes may now be packed alongside other buffers, for performance.
* `GpuMesh` has been renamed to `RenderMesh`, to reflect the fact that it no longer directly stores handles to GPU objects.
* Because meshes no longer have their own vertex and index buffers, the responsibility for the buffers has moved from `GpuMesh` (now called `RenderMesh`) to the `MeshAllocator` resource. To access the vertex data for a mesh, use `MeshAllocator::mesh_vertex_slice`. To access the index data for a mesh, use `MeshAllocator::mesh_index_slice`.

[`offset-allocator`]: https://github.com/pcwalton/offset-allocator

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+    Render, RenderApp, RenderSet,
+};
+
+/// A plugin that provides memory management services for mesh data.
```
```suggestion
/// A plugin that provides GPU buffer management for uploading [`Mesh`] data.
```

Something like this, maybe. Still not quite happy with it though.

pcwalton replied:
I reworded it.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+/// A plugin that provides memory management services for mesh data.
+pub struct MeshAllocatorPlugin;
+
+/// Manages the assignment of mesh data to hardware buffers.
```
```suggestion
/// Manages the assignment of mesh data to GPU buffers.
```

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+pub struct MeshAllocatorSettings {
+    /// The minimum size of a slab (hardware buffer), in bytes.
+    ///
+    /// The default value is 1 MB.
```
MB or MiB?

pcwalton replied:
I'm not inclined to indulge SI's attempt to redefine terms that everyone knows :)

JMS55 replied:
MiB is base 2 though, instead of base 10, which seems like a better fit for memory sizes to me :)

pcwalton replied:
MB is base 2 for everyone except SI. I don't want to indulge SI's attempt to redefine MB as base 10 for silly reasons.

pcwalton replied:
I don't like the use of MiB but I changed it anyway.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+    /// fragmentation but fewer expensive copy operations when growing the
+    /// buffer.
+    ///
+    /// The default value is 1.5.
```
Why not 2?

pcwalton replied:
See https://archive.ph/Z2R8w. tl;dr a value closer to the golden ratio increases the chance that if you reallocate more than once you're able to coalesce older spaces into the new space.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+/// [`MeshAllocatorSettings::large_threshold`]. Slabs are divided into *slots*,
+/// which are described in detail in the [`ElementLayout`] documentation.
+struct GeneralSlab {
+    /// The [`Allocator`] thatmanages the objects in this slab.
```
```suggestion
    /// The [`Allocator`] that manages the objects in this slab.
```

pcwalton replied:
Fixed

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+/// Slab objects are allocated in units of *slots*. Usually, each element takes
+/// up one slot, and so elements and slots are equivalent. Occasionally,
+/// however, a slot may consist of 2 or even 4 elements. This occurs when the
+/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`] (4). When we
```
```suggestion
/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`]. When we
```

Specifying 4 makes reading the sentence more confusing that not, imo.

pcwalton replied:
Removed

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+
+/// Holds information about all slabs scheduled to be allocated or reallocated.
+#[derive(Default, Deref, DerefMut)]
+struct SlabsToReallocate(HashMap<SlabId, SlabToReallocate>);
```
HashMap seems excessive here and in SlabToReallocate. Do we have some kind of SmallVec-like map type?

pcwalton replied:
No, we don't, and I don't think it's worth pulling a dependency and bloating Bevy's code size to optimize this. Reallocation is super slow no matter what. The overhead of a couple of CPU side allocations are a drop in the bucket next to allocating the buffer on GPU and copying over.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+    allocations_to_copy: HashMap<AssetId<Mesh>, SlabAllocation>,
+}
+
+impl Default for MeshAllocatorSettings {
```
Can you move this up to under the definition of MeshAllocatorSettings ?

pcwalton replied:
Done

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+                Render,
+                allocate_and_free_meshes
+                    .in_set(RenderSet::PrepareAssets)
+                    .before(prepare_assets::<GpuMesh>),
```
Why before here, and not after? This stands out as very weird to me.

pcwalton replied:
It's because `prepare_assets` clears the contents of the `ExtractedAssets<RenderMesh>` resource after processing it. Because `allocate_and_free_meshes` needs to use that data, it has to run first. (We don't want to merge `allocate_and_free_meshes` with `RenderMesh::prepare_asset` because, among other reasons, that would result in a lot of useless copying when a bunch of meshes load all at the same time and the buffer grows multiple times in a single frame, which happens a lot.)

JMS55 replied:
Am I correct that this this system runs on a 1-frame delay then?

pcwalton replied:
No. The order in a single frame is this:

1. `extract_render_asset` populates `ExtractedAssets<RenderMesh>` during the extract phase.
2. `allocate_and_free_meshes` examines the `ExtractedAssets<RenderMesh>` to perform the allocations during the asset prep phase.
3. `prepare_asset` processes the `ExtractedAssets<RenderMesh>` again, and clears it out.

So `extract_render_asset` produces the data, `allocate_and_free_meshes` inspects the data, and finally `prepare_asset` consumes the data. This pipeline happens entirely during the frame.

JMS55 replied:
Ah ok, thanks for the explanation. This makes sense to me, and mimics how I have it setup in my own code.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+
+    /// A generic function that copies either vertex or index data into a slab.
+    #[allow(clippy::too_many_arguments)]
+    fn copy_mesh_data(
```
```suggestion
    fn copy_element_data(
```

To match the terminology used elsewhere.

pcwalton replied:
Done

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+        return Cow::Borrowed(buffer);
+    }
+    let mut buffer = buffer.to_vec();
+    while buffer.len() % align != 0 {
```
You can just use the value of `align - (buffer.len() % align)` with `buffer.extend(iter::repeat(0).take(n))` here and avoid the while loop.

pcwalton replied:
That's wrong in the case in which the buffer is already of the right alignment. But if I wrap it in an if, sure.

pcwalton replied:
Oh wait, never mind, I have the if already above.

pcwalton replied:
Changed the line to your suggestion.

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -0,0 +1,1025 @@
+    Cow::Owned(buffer)
+}
+
+/// Returns either `"vertex "` or `"index "`, depending on the type of buffer
```
I'd rewrite this whole doc comment to something much simpler without the first paragraph. Maybe just like map BufferUsages to a string.

JMS55 commented on crates/bevy_render/src/render_asset.rs:
```diff
@@ -168,9 +168,9 @@ impl<A: RenderAsset> RenderAssetDependency for A {
-    extracted: Vec<(AssetId<A::SourceAsset>, A::SourceAsset)>,
-    removed: HashSet<AssetId<A::SourceAsset>>,
-    added: HashSet<AssetId<A::SourceAsset>>,
+    pub(crate) extracted: Vec<(AssetId<A::SourceAsset>, A::SourceAsset)>,
```
I'd like to make these pub in general.

pcwalton replied:
Done

JMS55 reviewed (changes requested):
Great docs as always. Looks generally good. If there any bugs or perf overhead from the allocator itself, we can iron them out later. I'm not confident that separate slabs for large buffers are needed, but we can try it out, I don't feel strongly about removing them either. Bevy really needs an example/benchmark for loading and unloading multiple assets in the background during gameplay.

The one thing I'm concerned about is that GpuMesh seems to be in an akward spot now, given that it doesn't actually hold any mesh data anymore. Not sure what we want to do with it now, I'll let @superdump weigh in.

I'd like to see:
* Perf test on WebGL2 to see how much of a hit the allocator logic is vs main
* A bit more detailed migration guide, specifically mentioning the GpuMesh change and how it now lives in MeshAllocator. Please mention the full types, not just the concepts.

BD103 commented:
The original PR was marked for release notes, so I'm going to add it here too!

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -119,6 +119,21 @@ pub struct MeshAllocatorSettings {
+impl Default for MeshAllocatorSettings {
+    fn default() -> Self {
+        Self {
+            // 1MB
```
This should be MiB since it's 1024 and not 1000 no?

pcwalton replied:
I don't like the use of MiB but I changed it anyway.

pcwalton commented on crates/bevy_render/src/mesh/mesh/mod.rs:
```diff
@@ -1665,7 +1661,6 @@ impl BaseMeshPipelineKey {
 #[derive(Debug, Clone)]
 pub struct GpuMesh {
```
I renamed it to `RenderMesh` since it's the render world version of a `Mesh`.

pcwalton commented on crates/bevy_render/src/mesh/mesh/mod.rs:
```diff
@@ -1665,7 +1661,6 @@ impl BaseMeshPipelineKey {
 #[derive(Debug, Clone)]
 pub struct GpuMesh {
     /// Contains all attribute data for each vertex.
```
Removed

pcwalton commented on crates/bevy_render/src/batching/gpu_preprocessing.rs:
```diff
@@ -186,7 +186,7 @@ pub struct IndirectParameters {
     pub instance_count: u32,
 
     /// The index of the first vertex we're to draw.
```
Reworked all comments here

JMS55 commented on crates/bevy_render/src/mesh/allocator.rs:
```diff
@@ -809,6 +807,9 @@ impl MeshAllocator {
         }
+
+        let command_buffer = encoder.finish();
+        render_queue.submit(iter::once(command_buffer));
```
You can also use [command_buffer]

pcwalton commented:
I've addressed all the review comments as much as I feel is appropriate. From running the examples on WebGL the performance seems fine. I can't do very in-depth performance analysis there without better diagnostics, and I don't think we should block this PR on better WebGL diagnostics.

tychedelia commented on examples/shader/shader_instancing.rs:
```diff
@@ -250,9 +256,11 @@ impl<P: PhaseItem> RenderCommand<P> for DrawMeshInstanced {
+        (meshes, render_mesh_instances, mesh_allocator): SystemParamItem<'w, '_, Self::Param>,
         pass: &mut TrackedRenderPass<'w>,
     ) -> RenderCommandResult {
+        let mesh_allocator = mesh_allocator.into_inner();
```
It might be nice to have a quick comment in the example pointing the user to further inspect the docs for `MeshAllocator` to learn more about why this is necessary.

pcwalton replied:
It's a borrow check workaround. I wrote a comment above it.

tychedelia reviewed (approved):
Looks good, thanks for the excellent documentation. Two small comments:
- It might be nice to have some logging to warn when a mesh spills over into the large object slab if it's near the boundary.
- We should follow up to rename the other `RenderAsset` e.g `GpuImage` -> `RenderImage  for consistency per the changes here.

NthTensor commented:
The issue with the bistro interior seems to have been resolved. I can replicate the opaque pass improvements. FPS did not improve in tests of low-end hardware (I am totally gpu bottlenecked on most scenes), but it also doesn't regress which is the main thing I was concerned about.

pcwalton commented:
The merge fallout is fixed. This is ready to be merged again.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1662 if you'd like to help out.
